--- ../src_base/minecraft/net/minecraft/dispenser/BehaviorDispenseMinecart.java
+++ ../src_work/minecraft/net/minecraft/dispenser/BehaviorDispenseMinecart.java
@@ -1,12 +1,16 @@
 package net.minecraft.dispenser;
 
+import net.minecraft.block.BlockDispenser;
 import net.minecraft.block.BlockRail;
 import net.minecraft.entity.item.EntityMinecart;
 import net.minecraft.item.ItemMinecart;
 import net.minecraft.item.ItemStack;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.EnumFacing;
+import net.minecraft.util.Vec3;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.block.BlockDispenseEvent;
 
 public class BehaviorDispenseMinecart extends BehaviorDefaultDispenseItem
 {
@@ -50,7 +54,18 @@
 
             var15 = -1.0D;
         }
-
+        BlockDispenseEvent ev = new BlockDispenseEvent(par1IBlockSource.getXInt(), par1IBlockSource.getYInt(), par1IBlockSource.getZInt(), par1IBlockSource.getWorld(), par2ItemStack, Vec3.vec3dPool.getVecFromPool(var5, var7 + var15, var9), this);
+        
+        if (!BlockDispenser.eventFired) {
+            if (MinecraftForge.EVENT_BUS.post(ev)) {
+                return par2ItemStack;
+            }
+            
+            if (!(ev.toDispense.getItem() instanceof ItemMinecart)) {
+                super.dispense(par1IBlockSource, ev.toDispense);
+                return par2ItemStack;
+            }
+        }
         EntityMinecart var17 = new EntityMinecart(var4, var5, var7 + var15, var9, ((ItemMinecart)par2ItemStack.getItem()).minecartType);
         var4.spawnEntityInWorld(var17);
         par2ItemStack.splitStack(1);
